### PR TITLE
More encoding/decoding error context

### PR DESCRIPTION
**Pull request by awly**
Make it easier to trace which field caused an error.

### CODE REVIEW

File: tpm2/structures.go [hunk 1]
```diff
@@ -228,20 +228,20 @@ func decodeECCParams(in *bytes.Buffer) (*ECCParams, error) {
...
-		return nil, err
+		return nil, fmt.Errorf("decoding Sign: %v", err)
 	}
 	if err := tpmutil.UnpackBuf(in, &params.CurveID); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding CurveID: %v", err)
 	}
 	if params.KDF, err = decodeKDFScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding KDF: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: "KDF" -> "KDFScheme"

File: tpm2/structures.go [hunk 2]
```diff
@@ -228,20 +228,20 @@ func decodeECCParams(in *bytes.Buffer) (*ECCParams, error) {
 	var err error
 
 	if params.Symmetric, err = decodeSymScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Symmetric: %v", err)
 	}
 	if params.Sign, err = decodeSigScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Sign: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: "Sign" -> "SigScheme"

File: tpm2/structures.go [hunk 3]
```diff
@@ -228,20 +228,20 @@ func decodeECCParams(in *bytes.Buffer) (*ECCParams, error) {
 	var err error
 
 	if params.Symmetric, err = decodeSymScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Symmetric: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: "Symmetric" -> "SymScheme"

File: tpm2/structures.go [hunk 4]
```diff
@@ -164,14 +164,14 @@ func decodeRSAParams(in *bytes.Buffer) (*RSAParams, error) {
...
 	}
 	if params.Sign, err = decodeSigScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Sign: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: "Sign" -> "SigScheme"

File: tpm2/structures.go [hunk 5]
```diff
@@ -164,14 +164,14 @@ func decodeRSAParams(in *bytes.Buffer) (*RSAParams, error) {
 	var err error
 
 	if params.Symmetric, err = decodeSymScheme(in); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Symmetric: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: "Symmetric" -> "SymScheme"

File: tpm2/structures.go [hunk 6]
```diff
@@ -265,13 +265,13 @@ func (s *SymScheme) encode() ([]byte, error) {
...
 	if err := tpmutil.UnpackBuf(in, &scheme.Alg); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding Alg: %v", err)
 	}
 	if scheme.Alg == AlgNull {
 		return nil, nil
 	}
 	if err := tpmutil.UnpackBuf(in, &scheme.KeyBits, &scheme.Mode); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("decoding KeyBigs, Mode: %v", err)
```
Comment (tmroeder) [inline tpm2/structures.go]: typo: "KeyBigs" -> "KeyBits"

Comment (awly) [reply tpm2/structures.go]: Done

**Comment by awly** [comment]
For all the other comments - I'm putting field name in the error, not type name. Does that sound reasonable?

**Comment by tmroeder** [comment]
Ah OK, if the message is about the name not the type, then that's fine.

**Pull request merged**